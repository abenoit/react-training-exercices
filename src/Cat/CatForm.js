import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Form, ControlLabel, FormControl, Button } from 'react-bootstrap';
import './Cat.css';

const propTypes = {
  addCat: PropTypes.func.isRequired
};

const defaultCat = {
  name: '',
  age: 1,
  meowsPerHour: 10,
  color: 'green'
};

class CatForm extends Component {
  constructor() {
    super();

    this.state = {
      cat: { ...defaultCat }
    };

    this.handleChangeName = this.handleChangeName.bind(this);
    this.handleChangeAge = this.handleChangeAge.bind(this);
    this.handleChangeMeowsPerHour = this.handleChangeMeowsPerHour.bind(this);
    this.handleChangeColor = this.handleChangeColor.bind(this);
    this.addCat = this.addCat.bind(this);
  }

  handleChangeName(event) {
    this.setState({ cat: { ...this.state.cat, name: event.target.value } });
  }

  handleChangeAge(event) {
    this.setState({
      cat: { ...this.state.cat, age: event.target.valueAsNumber }
    });
  }

  handleChangeMeowsPerHour(event) {
    this.setState({
      cat: { ...this.state.cat, meowsPerHour: event.target.valueAsNumber }
    });
  }

  handleChangeColor(event) {
    this.setState({
      cat: { ...this.state.cat, color: event.target.value }
    });
  }

  isSubmitDisabled() {
    return !this.state.cat.name || !this.state.cat.age;
  }

  addCat() {
    this.props.addCat(this.state.cat);
    this.setState({
      cat: { ...defaultCat }
    });
  }

  render() {
    return (
      <div className="CatForm">
        <h2> Add a new cat </h2>
        <Form inline>
          <ControlLabel>Name * </ControlLabel>
          <FormControl
            value={this.state.cat.name}
            placeholder="Name"
            onChange={this.handleChangeName}
          />

          <ControlLabel>Age * </ControlLabel>
          <FormControl
            type="number"
            value={this.state.cat.age}
            placeholder="Age"
            onChange={this.handleChangeAge}
          />

          <ControlLabel>MeowsPerHour</ControlLabel>
          <FormControl
            type="number"
            width="200px"
            value={this.state.cat.meowsPerHour}
            placeholder="MeowsPerHour"
            onChange={this.handleChangeMeowsPerHour}
          />

          <ControlLabel>Color</ControlLabel>
          <FormControl
            width="200px"
            value={this.state.cat.color}
            placeholder="Color"
            onChange={this.handleChangeColor}
          />

          <Button
            bsStyle="primary"
            onClick={this.addCat}
            disabled={this.isSubmitDisabled()}
          >
            Add this cat
          </Button>
        </Form>
      </div>
    );
  }
}

CatForm.propTypes = propTypes;
export default CatForm;
