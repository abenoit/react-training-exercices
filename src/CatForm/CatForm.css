.Cat {
  padding: 10px 0;
  font-size: 15px;
  display: flex;
  justify-content: center;
}

.Cat button {
  margin: 0 20px;
}

.Cat .info {
  text-align: left;
}

.Cat .name {
  font-size: 1.5em;
  padding: 0 5px;
  color: darkcyan;
}
