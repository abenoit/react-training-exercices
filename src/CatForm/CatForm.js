import { useState } from 'react';
import PropTypes from 'prop-types';
import { InputLabel, FormControl, Button, Input } from '@material-ui/core';
import './CatForm.css';

const propTypes = {
  addCat: PropTypes.func.isRequired,
};

const CatForm = (props) => {
  const [name, setName] = useState('');
  const [age, setAge] = useState(1);
  const [maxMeow, setMaxMeow] = useState(10);

  const isSubmitDisabled = () => {
    return !name || !age;
  };

  const addCat = () => {
    props.addCat({
      name,
      age,
      maxMeow,
    });

    setName('');
    setAge(1);
    setMaxMeow(10);
  };

  return (
    <div className="CatForm">
      <h2> Add a new cat </h2>
      <form>
        <FormControl>
          <InputLabel>Name * </InputLabel>
          <Input
            value={name}
            placeholder="Name"
            onChange={(evt) => setName(evt.target.value)}
          />
        </FormControl>

        <FormControl>
          <InputLabel>Age * </InputLabel>
          <Input
            type="number"
            value={age}
            placeholder="Age"
            onChange={(evt) => setAge(evt.target.value)}
          />
        </FormControl>

        <FormControl>
          <InputLabel>MeowsPerHour</InputLabel>
          <Input
            type="number"
            width="200px"
            value={maxMeow}
            placeholder="Max meows"
            onChange={(evt) => setMaxMeow(evt.target.value)}
          />
        </FormControl>

        <Button
          variant="contained"
          bsStyle="primary"
          onClick={addCat}
          disabled={isSubmitDisabled()}
        >
          Add this cat
        </Button>
      </form>
    </div>
  );
};

CatForm.propTypes = propTypes;
export default CatForm;
