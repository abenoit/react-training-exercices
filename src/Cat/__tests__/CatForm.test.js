import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import CatForm from '../CatForm';

describe('CatForm', () => {
  it('should display a cat form', () => {
    const catForm = renderer.create(<CatForm addCat={() => {}} />).toJSON();
    expect(catForm).toMatchSnapshot();
  });

  describe('handleChangeName', () => {
    test("should update cat's name in state", () => {
      const catForm = shallow(<CatForm addCat={() => {}} />);
      const event = { target: { value: 'new name' } };

      catForm.instance().handleChangeName(event);

      expect(catForm.state('cat').name).toEqual('new name');
    });
  });

  describe('handleChangeAge', () => {
    test("should update cat's age in state", () => {
      const catForm = shallow(<CatForm addCat={() => {}} />);
      const event = { target: { valueAsNumber: 2 } };

      catForm.instance().handleChangeAge(event);

      expect(catForm.state('cat').age).toEqual(2);
    });
  });

  describe('handleChangeMeowsPerHour', () => {
    test("should update cat's meowsPerHour in state", () => {
      const catForm = shallow(<CatForm addCat={() => {}} />);
      const event = { target: { valueAsNumber: 4 } };

      catForm.instance().handleChangeMeowsPerHour(event);

      expect(catForm.state('cat').meowsPerHour).toEqual(4);
    });
  });

  describe('handleChangeColor', () => {
    test("should update cat's color in state", () => {
      const catForm = shallow(<CatForm addCat={() => {}} />);
      const event = { target: { value: 'pink' } };

      catForm.instance().handleChangeColor(event);

      expect(catForm.state('cat').color).toEqual('pink');
    });
  });

  describe('handleChangeColor', () => {
    it('should return false if name and age are set', () => {
      const catForm = shallow(<CatForm addCat={() => {}} />);
      catForm.setState({ cat: { name: 'lala', age: 3 } });

      const isSubmitDisabled = catForm.instance().isSubmitDisabled();

      expect(isSubmitDisabled).toBeFalsy();
    });

    it('should return true if name is not set', () => {
      const catForm = shallow(<CatForm addCat={() => {}} />);
      catForm.setState({ cat: { name: '', age: 3 } });

      const isSubmitDisabled = catForm.instance().isSubmitDisabled();

      expect(isSubmitDisabled).toBeTruthy();
    });

    it('should return true if age is not set', () => {
      const catForm = shallow(<CatForm addCat={() => {}} />);
      catForm.setState({ cat: { name: 'lala' } });

      const isSubmitDisabled = catForm.instance().isSubmitDisabled();

      expect(isSubmitDisabled).toBeTruthy();
    });
  });

  describe('addCat', () => {
    it('should call callback in props', () => {
      const catForm = shallow(<CatForm addCat={jest.fn()} />);

      catForm.instance().addCat();

      expect(catForm.instance().props.addCat).toHaveBeenCalled();
    });

    it('should reinit the state', () => {
      const catForm = shallow(<CatForm addCat={jest.fn()} />);

      catForm.instance().addCat();

      expect(catForm.state('cat').name).toEqual('');
      expect(catForm.state('cat').age).toEqual(1);
      expect(catForm.state('cat').meowsPerHour).toEqual(10);
      expect(catForm.state('cat').color).toEqual('green');
    });
  });
});
