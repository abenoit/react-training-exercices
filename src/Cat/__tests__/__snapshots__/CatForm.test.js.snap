// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CatForm should display a cat form 1`] = `
<div
  className="CatForm"
>
  <h2>
     Add a new cat 
  </h2>
  <form
    className="form-inline"
  >
    <label
      className="control-label"
      htmlFor={undefined}
    >
      Name * 
    </label>
    <input
      className="form-control"
      id={undefined}
      onChange={[Function]}
      placeholder="Name"
      type={undefined}
      value=""
    />
    <label
      className="control-label"
      htmlFor={undefined}
    >
      Age * 
    </label>
    <input
      className="form-control"
      id={undefined}
      onChange={[Function]}
      placeholder="Age"
      type="number"
      value={1}
    />
    <label
      className="control-label"
      htmlFor={undefined}
    >
      MeowsPerHour
    </label>
    <input
      className="form-control"
      id={undefined}
      onChange={[Function]}
      placeholder="MeowsPerHour"
      type="number"
      value={10}
      width="200px"
    />
    <label
      className="control-label"
      htmlFor={undefined}
    >
      Color
    </label>
    <input
      className="form-control"
      id={undefined}
      onChange={[Function]}
      placeholder="Color"
      type={undefined}
      value="green"
      width="200px"
    />
    <button
      className="btn btn-primary"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      Add this cat
    </button>
  </form>
</div>
`;
